File: data_structures/python/hash_map.py
from typing import List


class HashNode:
    def __init__(self):
        self.key = None
        self.val = None
        self.next: HashNode = None


class HashMap:

    def __init__(self):
        self._storage: List[HashNode] = []
        self._capacity = 0
        self._limit = 0  # When the capacity reaches this, then resize the array.
        self._size = 0
        self.resize(8)

    def resize(self, new_size: int):
        old_capacity = self._capacity

        # Increase the capacity.
        new_storage = []
        for _ in range(new_size):
            new_storage.append(HashNode())

        self._capacity = new_size
        self._limit = int(self._capacity * 0.7)

        # Move the items.
        for i in range(old_capacity):
            node = self._storage[i]
            while node.key is not None:
                self._set_item_to_array(node.key, node.val, new_storage)
                node = node.next

        # Set the storage to the new.
        self._storage = new_storage

    @staticmethod
    def get_hash_index(key, capacity):
        return hash(key) % capacity

    def __getitem__(self, key):
        index = self.get_hash_index(key, len(self._storage))
        node = self._storage[index]
        while node.key != key:
            node = node.next
            if node is None:
                raise Exception(f"Key {key} not found in Hash Table.")
        return node.val

    def __setitem__(self, key, value):

        if key is None:
            raise Exception("Cannot use None for key.")
        size_increase = self._set_item_to_array(key, value, self._storage)

        self._size += size_increase
        if self._size >= self._limit:
            self.resize(int(self._capacity * 1.5))

    def _set_item_to_array(self, key, value, array):
        index = self.get_hash_index(key, len(array))
        node = array[index]

        while node.key is not None and node.key != key:
            node = node.next

        node.key = key
        node.val = value
        if node.next is None:
            node.next = HashNode()
            return 1
        else:
            return 0

    def size(self):
        return self._size

    def display(self):

        for i in range(self._capacity):
            display_line = [f"{str(i).zfill(3)}: ["]
            node = self._storage[i]
            while node.key is not None:
                display_line.append(f"{node.key} ({node.val})")

                if node.next.key is not None:
                    display_line.append(" -> ")

                node = node.next

            display_line.append("]")
            print("".join(display_line))

        print(f"\nCapacity: {self._capacity}")
        print(f"Size: {self._size}")
        print(f"Limit: {self._limit}")

File: data_structures/go/hash_map.go
package ds

import (
	"fmt"
	"hash/fnv"
)

type HashNode struct {
	key      string
	val      interface{}
	next     *HashNode
	occupied bool
}

type HashMap struct {
	storage  []*HashNode
	size     int
	capacity int
	limit    int
}

func NewHashMap() HashMap {
	hashMap := HashMap{
		storage:  make([]*HashNode, 0),
		size:     0,
		capacity: 0,
		limit:    0,
	}
	hashMap.resize(8)
	return hashMap
}

func getHashIndex(key string, capacity int) int {
	algorithm := fnv.New32a()
	algorithm.Write([]byte(key))
	return int(algorithm.Sum32()) % capacity
}

func (hm *HashMap) Set(key string, value interface{}) {

	hm.size += setInternal(key, value, hm.capacity, hm.storage)
	if hm.size >= hm.limit {
		hm.resize(hm.capacity * 2)
	}
}

func setInternal(key string, value interface{}, capacity int, nodes []*HashNode) int {
	index := getHashIndex(key, capacity)
	node := nodes[index]

	for node.occupied && node.key != key {
		node = node.next
	}

	node.key = key
	node.val = value

	if !node.occupied {
		newNode := HashNode{occupied: false}
		node.next = &newNode
		node.occupied = true
		return 1
	}

	return 0
}

func (hm *HashMap) Get(key string) interface{} {
	index := getHashIndex(key, hm.capacity)
	node := hm.storage[index]

	for node.occupied && node.key != key {
		node = node.next
	}

	if !node.occupied {
		panic("Key not found!")
	} else {
		return node.val
	}
}

func (hm *HashMap) Display() {

	for i := 0; i < hm.capacity; i++ {
		node := hm.storage[i]
		fmt.Printf("%03v: [", i)
		for node.occupied {
			fmt.Printf("%v (%v)", node.key, node.val)
			node = node.next
			if node.occupied {
				fmt.Printf(" -> ")
			}
		}
		fmt.Printf("]\n")
	}
	fmt.Printf("\nSize: %v\n", hm.size)
	fmt.Printf("Capacity: %v\n", hm.capacity)
	fmt.Printf("Limit: %v\n", hm.limit)
}

func (hm *HashMap) Size() int {
	return hm.size
}

func (hm *HashMap) resize(newSize int) {
	// Create the new slice.
	newStorage := make([]*HashNode, newSize)
	for i := 0; i < newSize; i++ {
		node := HashNode{occupied: false}
		newStorage[i] = &node
	}

	// Transfer the data.
	for i := 0; i < hm.capacity; i++ {
		node := hm.storage[i]
		for node.occupied {
			setInternal(node.key, node.val, newSize, newStorage)
			node = node.next
		}
	}

	// Set the references.
	hm.capacity = newSize
	hm.limit = hm.capacity
	hm.storage = newStorage
}
